benchmarks/lockhammer/src/lockhammer.c

```diff
@@ -271,6 +326,13 @@ void* hmr(void *ptr)
 
         /* Spin until the "marshal" sets the appropriate bit */
         wait64(&sync_lock, (nthrds * 2) | 1);
+
+        //TODO: Need more extensive testing on whether we need
+        //      multiple calibration loops to run.
+        //CALIBRATE_TIMER();
```

Comment:
Unless I'm misreading then without this enabled in this branch I don't think hold and post will be calibrated for any threads other than the one scheduled on core 0 since hold_count and post_count which CALIBRATE_TIMER modifies are per-thread stack variables.

Also this is good but ideally it'd be better to do add a second synchronized start of all the threads for the calibration loop in addition to the existing synchronized start for the lock test or else it's possible that the calibration loops misalign and incorrectly calibrate for hyperthreaded systems (ie, if the calibration code skews then the first x nanoseconds will be single thread per core, the next y nanoseconds two thread per core with ~1/2 the throughput, and the last z nanoseconds again one thread per core).  I'm not really sure how to make that work given that all of the cores would then be doing binary searches of their own timer loops and actually based on how things get scheduled it's theoretically possible to livelock if one core continually flip/flops between running one or two threads at the same time and bounces back and forth between thinking it needs more or fewer loops on subsequent runs.

Reply:
Correct, right now I am only measuring 1 thread, I am seeing measurements with all threads get stuck sometimes.

That is a good point about getting stuck because the test loops are not synchronized. In fact I was seeing something to that extent while testing this yesterday on some multi-threaded enabled machines.  Dynamic frequency scaling is also leading to issues here, meaning the overhead calculations for loop measurement need to be redone more frequently, otherwise I see the calibration get stuck thinking it is continually measuring an outlier.

I'll give this code another go-other later this week and hopefully clean up these issues.  Thanks for the feedback.  I believe you are right in that all the threads need to do their search in a synchronized manner (or at least all the threads in the same core need to).